### DISCUSSION TITLE
is possible to implement non blocking wlan scan (especially on esp32 port)?

### DISCUSSION THREAD

**Discussion by puppet13th** [micropython/micropython#9828] [Core Development] [answered]
Hello everybody.
currently sta.scan is blocking function, is it possible to implement it non blocking ways like sta.connect?

**Comment by stephanelsmith** [comment, marked as answer]
I don't think so.  I would personally love to see callbacks on every scan result, similar to the bluetooth IRQ callback system.  Which could then be wrapped into an asyncio system.

**Reply by puppet13th** [reply]
That one will be nice too, can incorporate asyncio to handle scan result.
Eg like polling scan for known wlan ssid, and connect known network when in range( show up in scan result)

**Comment by jonathanhogg** [comment]
I'm a bit late to this question, but in case it's still useful to you (or useful to someone else) I wrap the `.scan()` method in a thread and that works fine on ESP32.

Adapted excerpt from my code:

```python
class WifiManager:
    def __init__(self):
        self.wlan = network.WLAN(network.STA_IF)
        self.wlan.active(True)
        self._scan_start = _thread.allocate_lock()
        self._scan_start.acquire()
        self._scan_complete = asyncio.ThreadSafeFlag()
        self._scan_thread = _thread.start_new_thread(self._scan, ())
        self._scan_results = None

    def _scan(self):
        while True:
            self._scan_start.acquire()
            try:
                self._scan_results = self.wlan.scan()
            except Exception as exc:
                Log.warning("Exception running scan", exc=exc)
                self._scan_results = []
            self._scan_complete.set()

    async def scan(self):
        self._scan_start.release()
        await self._scan_complete.wait()
        self._scan_complete.clear()
        return self._scan_results
```

I keep one, long-running background thread to look after scans. This thread is woken up by releasing a `_thread.Lock` and then notifies it is finished via an `asyncio.ThreadSafeFlag` so that the scan results can be `await`ed.

**Reply by puppet13th** [reply]
@jonathanhogg thank your for your reply. does other coro/task still running while waiting for scan result?

**Reply by jonathanhogg** [reply]
Yes. The `await self._scan_complete.wait()` explicitly hands control back to the `asyncio` scheduler until the scan, which is running on a background thread, has completed.

This approach has been rock solid for me, though I would certainly try to avoid touching the `network.WLAN()` object in any other way while the scan is running in case.

**Reply by stephanelsmith** [reply]
Wow, just seeing this again, incredible, can't way to try it.  Thank you for the tip.  I hadn't considered using threads. *bump*
